### ISSUE TITLE
Support external schema reference

### ISSUE THREAD

**Issue by MugeSo** [kogosoftwarellc/express-openapi#17]
## Problem
Now my api receives json schema object.

so, i wrote:
```javascript
// in handler
post.apiDoc = {
    description: "create new schema definition",
    operationId: "createSchema",
    parameters: [
        {
            in: 'body',
            name: 'schema',
            schema: {$ref: '#/definitions/schema' }
        }
    ],
    responses: {
        "201": { $ref: '#/responses/created' },
        "default": { $ref: "#/responses/error" }
    }
}

// in apidoc.js#definitions
{ 
    "schema": { "$ref": "http://json-schema.org/schema" }
}  
```

Then i sent request to it and got:
```json
{
    "status":400,
    "errors":[
        {
            "message":"no such schema <http://json-schema.org/schema>",
            "schema":{"$ref":"http://json-schema.org/schema"},
            "location":"body"
        }
    ]
}
```

## Suggestion

1.  support external schema
2.  support maping external id to local file (for when can not connect to server hosting the external schema or the url don't serve schema json.)
3.  keep external reference as is in exposed api doc, not expanded loaded schema.

**Comment by MugeSo** [comment]
For SRP, local file loading should be omitted.
Just requires passing {id : SchemaObject} hash to ‘jsonschema` via `express-openapi-validation` or `express-openapi-response-validation`.
And also, of course, I want to pass it via initialize argument of express-openapi.

**Comment by jsdevel** [comment]
Can you try this?

```
$ref: 'http://json-schema.org/draft-04/schema#'
```

**Comment by MugeSo** [comment]
I got the same error :cry:

**Comment by jsdevel** [comment]
@MugeSo this will likely take a decent amount of work.  Have a look at https://github.com/tdegrunt/jsonschema#dereferencing-schemas.  As you can see, we'd need to update each module that uses `jsonschema`.

Given that, one of the designs of openapi is to be a limited subset of jsonschema.  I'm curious to know what intentions you have in referencing `http://json-schema.org/schema`?

**Comment by MugeSo** [comment]
@jsdevel 
At first, i know i should use 'http://json-schema.org/draft-04/schema#', 'http://json-schema.org/schema' is my mistake.

The reason why I reference it is to contain external defined schema in api-doc is nonsense.
When api-doc contains them, it get large wastefully.

And as i said in https://github.com/kogosoftwarellc/express-openapi/issues/17#issuecomment-196648659 , i don't need dynamic loading.
So, `deferencing-schemas` is not needed. just call `v.addSchema`. 

Like:
```javascript
Object.keys(externalSchemas).forEach(function(key) {
    v.addSchema(schemas[key], key);
});
```

If you want, I send PR to each modules :wink:

**Comment by jsdevel** [comment]
PRs are welcomed!

**Comment by MugeSo** [comment]
@jsdevel 
I fond `$ref` of parameters object are also unresolved.
e.g:

```javascript
parameters = [
   { $ref: '#/parameters/myparameter' }
]
```
And I considered about it and this again.

Now I think it's better to use [json-refs](https://www.npmjs.com/package/json-refs) or something to resolve any references(not only schema) in api doc.

How do you about?

**Comment by jsdevel** [comment]
So `$ref` actually isn't listed as a valid property in `parameters`.  See the [spec](https://github.com/OAI/OpenAPI-Specification/blob/master/schemas/v2.0/schema.json#L607).

**Comment by jsdevel** [comment]
I usually use code to overcome this shortcoming of openapi/swagger

```javascript
parameters: [
  myParameterModule.myParameter
]
```

`myParameterModule` looks like this:

```javascript
  myParameter: {
    name: 'foo',
    type: 'string',
    in: 'query'
  }
```

**Comment by MugeSo** [comment]
`$ref` is valid.

See `parameters` property.
https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md#operationObject
https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md#path-item-object

https://github.com/OAI/OpenAPI-Specification/blob/master/schemas/v2.0/schema.json#L1482

**Comment by jsdevel** [comment]
I see it: https://github.com/OAI/OpenAPI-Specification/blob/master/schemas/v2.0/schema.json#L1482

Then it should work.  Have you tried this?

```javascript
parameters = [
   { $ref: '#/definitions/myparameter' }
]
```

**Comment by MugeSo** [comment]
It's wrong. `parameters` property of `Swagger Object` is for define parameters.
See:
https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md#fixed-fields
https://github.com/OAI/OpenAPI-Specification/blob/master/schemas/v2.0/schema.json#L186

**Comment by MugeSo** [comment]
And it doesn't work because `openapi-jsonschema-parameters` doesn't handle that.

**Comment by jsdevel** [comment]
@MugeSo can you open a PR to address this?  This is a good find.

**Comment by MugeSo** [comment]
Also `resposes` property of `Operation Object'.
https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md#responses-object

> can you open a PR to address this? This is a good find.

I'll try it.
But it takes some time because it seems complex.
(And it's 3:25am here)

**Comment by jsdevel** [comment]
Go to bed! lol

**Comment by jsdevel** [comment]
Yea, I feel this would be a major addition to the framework.  Perhaps we should track responses and parameters separately?

**Comment by jsdevel** [comment]
[json-ref-lite](https://github.com/coderofsalvation/json-ref-lite) looks pretty cool too.

**Comment by MugeSo** [comment]
I've seen `json-ref-lite`.
It' nice except that it resolve reference start with `http:` only through network.
This causes problem when schema host server is down.

However, unfortunately, other library i found is async.
If we async one, `openapi.initialize` get async too :(

**Comment by MugeSo** [comment]
OTTOMH
```js
var jsonRef = requrie('json-ref-lite');

var externalSchemas = args.externalSchemas || [];
// create id base reference map
var ids = jsonRef.findId(externalSchemas.concat(apiDoc.definitions));
var resolved = copy(methodDoc);// json-ref-lite is destractive
jsonRef.replace(resolved, ids, apiDoc);
```

**Comment by MugeSo** [comment]
The reason why use only `apiDoc.definitions` is [Reference Object](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md#responses-object) supports only canonical reference.

`definitions` is array of [JSON Schema Object](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md#schemaObject) and it has own `$ref` definition in spec.

> $ref - As a JSON Reference

**Comment by MugeSo** [comment]
Sorry, in the previous code, `ids` include all ids in externalSchemas.
And external thing not only schema.

```js
var jsonRef = requrie('json-ref-lite');

// id -> json object map
var externalJsons = args.externalJsons || {};

// create id based reference map
var ids = sallowMerge(jsonRef.findId(apiDoc.definitions), externalSchemas);
var resolvedMethodDoc = copy(methodDoc);// json-ref-lite is destractive
jsonRef.replace(resolvedMethodDoc, ids, apiDoc);
```

**Comment by MugeSo** [comment]
Finally:
```js
// id -> json object map
jsonRef.cache = copy(args.externalJsons) || {};

var resolvedMethodDoc = jsonRef.replace(
  copy(methodDoc), // json-ref-lite is destractive
 jsonRef.findId(apiDoc.definitions),
 apiDoc
);
```

OK, I send PR soon.

**Comment by MugeSo** [comment]
Unfortunately, current implementation works with invalid reference like:
```js
{
  responses: {
    200: {
      $ref: '#/definitions/User'
    },
    default: {
      $ref: '#/definitions/Error'
    }
  }
};
```

This should be:
```js
{
  responses: {
    200: {
      $ref: '#/responses/User'
    },
    default: {
      $ref: '#/responses/Error'
    }
  }
};

// with apiDoc
{
  responses: {
    User: {
        description: "Response for user representation",
        schema: { $ref: "#/definitions/User" }
    },
    Error: {
        description: "Generic error response",
        schema: { $ref: "#/definitions/Error" }
  }
}
```

See: https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md#responses-object

So, what should I do? Can I cause BC break?
If so, it need more time because i need fix some tests existing.

**Comment by jsdevel** [comment]
Let me take a crack at this.  After thinking about it some more, I feel I have a simple solution.

**Comment by MugeSo** [comment]
I found simple solution too.

```js
function wrongReferenceFixed(responses) {
  var fixed = copy(responses);
  for(statusCode in responses) {
    if (responses[statusCode].$ref && /^#\/definitions\//.testr(esponses[statusCode].$ref) {
        responses[statusCode] = {
            description: statusCode + " Response",
            schema: responses[statusCode]
        };
    }
  }
}
```

**Comment by jsdevel** [comment]
@MugeSo can you review #20 ?

**Comment by jsdevel** [comment]
@MugeSo with version `v0.16.0`, `parameters` and `responses` have support for local `$ref`s whereas before they didn't.  If you can submit a PR for external refs then that would be great.  Thanks for all your contributions!

**Comment by jsdevel** [comment]
@MugeSo after looking into `json-ref-lite` a bit more, I have a few things about it that I'm not sure would be good for `express-openapi`:

* It's relying on `spawnSync` which is available in the latest versions of node, but isn't in older versions.
* When `spawnSync` isn't available, it tries to install a native add on.  I'd really like to avoid native add ons if at all possible.
* Initialization time will increase with the added latency.

It may be better to update `express-openapi-validation` and `express-openapi-response-validation` to use `jsonschema`'s async loading features for external schemas.

**Comment by MugeSo** [comment]
`spawnSync' is used in `sync-request` right?
But it's just optional.
No network access is needed when jsonRef.cache has every external JSONs.

[My solution](https://github.com/kogosoftwarellc/express-openapi/issues/17#issuecomment-197658250) do that.

**Comment by MugeSo** [comment]
After https://github.com/coderofsalvation/json-ref-lite/pull/2 is merged i'll send PR ;)

**Comment by jsdevel** [comment]
Closed by #30 .
